File: app/lib/models/filter_schema.dart
import 'package:json_annotation/json_annotation.dart';

part 'filter_schema.g.dart';

/// Type of parameter value.
enum ParameterType {
  @JsonValue('boolean')
  boolean,
  @JsonValue('integer')
  integer,
  @JsonValue('number')
  number,
  @JsonValue('string')
  string,
  @JsonValue('enum')
  enumType,
}

/// Type of UI widget to render for a parameter.
enum WidgetType {
  @JsonValue('slider')
  slider,
  @JsonValue('dropdown')
  dropdown,
  @JsonValue('checkbox')
  checkbox,
  @JsonValue('textfield')
  textfield,
  @JsonValue('number')
  number,

  /// A path, with a Browse button beside the field. Never inferred — a string
  /// parameter still defaults to [textfield], so a schema opts in explicitly.
  @JsonValue('filepicker')
  filepicker,
}

/// VapourSynth-specific parameter configuration.
@JsonSerializable()
class VapourSynthConfig {
  /// The parameter name in VapourSynth (may differ from schema name).
  final String name;

  const VapourSynthConfig({required this.name});

  factory VapourSynthConfig.fromJson(Map<String, dynamic> json) =>
      _$VapourSynthConfigFromJson(json);
  Map<String, dynamic> toJson() => _$VapourSynthConfigToJson(this);
}

/// UI configuration for a parameter.
@JsonSerializable()
class ParameterUiConfig {
  /// Display label for the parameter.
  final String? label;

  /// Description/tooltip for the parameter.
  final String? description;

  /// Type of widget to render.
  final WidgetType? widget;

  /// Decimal precision for number display.
  final int? precision;

  /// Whether this parameter is hidden from the UI.
  final bool? hidden;

  /// Condition for when this parameter is visible.
  /// Keys are parameter names, values are the required values (or list of values).
  final Map<String, dynamic>? visibleWhen;

  /// Custom labels for boolean values (renders as dropdown instead of toggle).
  /// Example: {"true": "Top Field First", "false": "Bottom Field First"}
  final Map<String, String>? booleanLabels;

  /// Extensions the file picker offers, without dots — `["srt", "ass"]`.
  /// Only meaningful for [WidgetType.filepicker]; null means any file.
  final List<String>? fileExtensions;

  const ParameterUiConfig({
    this.label,
    this.description,
    this.widget,
    this.precision,
    this.hidden,
    this.visibleWhen,
    this.booleanLabels,
    this.fileExtensions,
  });

  factory ParameterUiConfig.fromJson(Map<String, dynamic> json) =>
      _$ParameterUiConfigFromJson(json);
  Map<String, dynamic> toJson() => _$ParameterUiConfigToJson(this);
}

/// Definition of a single parameter.
@JsonSerializable()
class ParameterDefinition {
  /// Type of this parameter.
  final ParameterType type;

  /// Default value (type depends on [type]).
  @JsonKey(name: 'default')
  final dynamic defaultValue;

  /// Minimum value (for number/integer types).
  final double? min;

  /// Maximum value (for number/integer types).
  final double? max;

  /// Step size for sliders (for number/integer types).
  final double? step;

  /// Available options (for enum type).
  final List<String>? options;

  /// VapourSynth-specific configuration.
  final VapourSynthConfig? vapoursynth;

  /// UI configuration.
  final ParameterUiConfig? ui;

  /// Whether this parameter can be enabled/disabled with a checkbox.
  /// When disabled (null value), the parameter is not passed to VapourSynth.
  final bool? optional;

  const ParameterDefinition({
    required this.type,
    required this.defaultValue,
    this.min,
    this.max,
    this.step,
    this.options,
    this.vapoursynth,
    this.ui,
    this.optional,
  });

  /// Get the VapourSynth parameter name (falls back to schema name if not specified).
  String getVsName(String schemaName) => vapoursynth?.name ?? schemaName;

  /// Check if value is within valid range.
  bool isValidValue(dynamic value) {
    if (value == null) return true;

    switch (type) {
      case ParameterType.boolean:
        return value is bool;
      case ParameterType.integer:
        if (value is! int) return false;
        if (min != null && value < min!) return false;
        if (max != null && value > max!) return false;
        return true;
      case ParameterType.number:
        if (value is! num) return false;
        if (min != null && value < min!) return false;
        if (max != null && value > max!) return false;
        return true;
      case ParameterType.string:
        return value is String;
      case ParameterType.enumType:
        return value is String && (options?.contains(value) ?? true);
    }
  }

  factory ParameterDefinition.fromJson(Map<String, dynamic> json) =>
      _$ParameterDefinitionFromJson(json);
  Map<String, dynamic> toJson() => _$ParameterDefinitionToJson(this);
}

/// Definition of a filter method (e.g., DeHalo_alpha, YAHR).
@JsonSerializable()
class MethodDefinition {
  /// Unique identifier for this method.
  final String id;

  /// Display name.
  final String name;

  /// Description of what this method does.
  final String? description;

  /// VapourSynth function to call (e.g., "haf.DeHalo_alpha").
  final String function;

  /// List of parameter IDs that this method uses.
  final List<String> parameters;

  /// Whether this method only appears in advanced mode.
  ///
  /// Lets one filter offer a short, curated method list to everyone and the
  /// full set to someone who has asked for it — the mechanism that keeps a
  /// "Noise Reduction" row from becoming a dropdown of sixteen names. A method
  /// the user has actually selected is always shown regardless, so a preset or
  /// saved job never silently loses its method; see
  /// [FilterSchema.visibleMethods].
  final bool advancedOnly;

  const MethodDefinition({
    required this.id,
    required this.name,
    this.description,
    required this.function,
    required this.parameters,
    this.advancedOnly = false,
  });

  factory MethodDefinition.fromJson(Map<String, dynamic> json) =>
      _$MethodDefinitionFromJson(json);
  Map<String, dynamic> toJson() => _$MethodDefinitionToJson(this);
}

/// One VapourSynth call a filter makes, for the "what is this actually doing"
/// readout in advanced mode.
///
/// Most filters need none of this: a method already names its own [function],
/// and the readout uses that. This exists for the passes that are not a single
/// call — Colour Correction alone can invoke six different things depending on
/// which of its switches are on — where naming one function would be a lie and
/// naming none is what prompted the request.
///
/// [activeWhen] is evaluated against the filter's current parameter values with
/// exactly the same matcher as `visibleWhen`, so the readout can distinguish
/// what is running now from what the pass is merely capable of. An entry with
/// no condition is always active.
@JsonSerializable()
class ImplementationEntry {
  /// The call as it appears in the generated script, e.g. `core.retinex.MSRCP`.
  final String function;

  /// What this call contributes, in the user's vocabulary rather than the
  /// plugin's — "shadow detail", not "multi-scale retinex".
  final String? role;

  /// Condition under which this call is actually made. Null means always.
  final Map<String, dynamic>? activeWhen;

  const ImplementationEntry({
    required this.function,
    this.role,
    this.activeWhen,
  });

  factory ImplementationEntry.fromJson(Map<String, dynamic> json) =>
      _$ImplementationEntryFromJson(json);
  Map<String, dynamic> toJson() => _$ImplementationEntryToJson(this);
}

/// UI section grouping parameters together.
@JsonSerializable()
class UiSection {
  /// Section title.
  final String title;

  /// Parameter IDs in this section.
  final List<String> parameters;

  /// Whether the section is expanded by default.
  final bool expanded;

  /// Whether this section only appears in advanced mode.
  final bool advancedOnly;

  const UiSection({
    required this.title,
    required this.parameters,
    this.expanded = true,
    this.advancedOnly = false,
  });

  factory UiSection.fromJson(Map<String, dynamic> json) =>
      _$UiSectionFromJson(json);
  Map<String, dynamic> toJson() => _$UiSectionToJson(this);
}

/// UI layout configuration.
@JsonSerializable()
class UiLayout {
  /// Sections to organize parameters.
  final List<UiSection>? sections;

  const UiLayout({this.sections});

  factory UiLayout.fromJson(Map<String, dynamic> json) =>
      _$UiLayoutFromJson(json);
  Map<String, dynamic> toJson() => _$UiLayoutToJson(this);
}

/// Dependencies required by a filter.
@JsonSerializable()
class FilterDependencies {
  /// Python packages required (e.g., ["havsfunc", "mvsfunc"]).
  final List<String>? plugins;

  /// VapourSynth plugins required (e.g., ["libmvtools.dll"]).
  @JsonKey(name: 'vs_plugins')
  final List<String>? vsPlugins;

  /// Optional plugins that enable additional features.
  final List<String>? optional;

  const FilterDependencies({
    this.plugins,
    this.vsPlugins,
    this.optional,
  });

  factory FilterDependencies.fromJson(Map<String, dynamic> json) =>
      _$FilterDependenciesFromJson(json);
  Map<String, dynamic> toJson() => _$FilterDependenciesToJson(this);
}

/// A parameter preset group - a dropdown that sets one or more parameters.
@JsonSerializable()
class ParameterPreset {
  /// Display label for the preset selector.
  final String label;

  /// Optional description.
  final String? description;

  /// Default option key.
  @JsonKey(name: 'default')
  final String? defaultOption;

  /// Map of option display name to parameter values to apply.
  final Map<String, Map<String, dynamic>> options;

  /// Condition for when this preset is visible.
  final Map<String, dynamic>? visibleWhen;

  const ParameterPreset({
    required this.label,
    this.description,
    this.defaultOption,
    required this.options,
    this.visibleWhen,
  });

  factory ParameterPreset.fromJson(Map<String, dynamic> json) =>
      _$ParameterPresetFromJson(json);
  Map<String, dynamic> toJson() => _$ParameterPresetToJson(this);
}

/// Code generation configuration.
@JsonSerializable()
class CodeTemplate {
  /// Import statements needed.
  final List<String>? imports;

  /// How to generate code: "method" (use method.function) or custom template.
  final String? generate;

  /// Custom code body template (if generate != "method").
  final String? body;

  const CodeTemplate({
    this.imports,
    this.generate,
    this.body,
  });

  factory CodeTemplate.fromJson(Map<String, dynamic> json) =>
      _$CodeTemplateFromJson(json);
  Map<String, dynamic> toJson() => _$CodeTemplateToJson(this);
}

/// Preset configuration for filters with predefined parameter sets.
@JsonSerializable()
class PresetDefinition {
  /// Preset name.
  final String name;

  /// Parameter values for this preset.
  final Map<String, dynamic> values;

  const PresetDefinition({
    required this.name,
    required this.values,
  });

  factory PresetDefinition.fromJson(Map<String, dynamic> json) =>
      _$PresetDefinitionFromJson(json);
  Map<String, dynamic> toJson() => _$PresetDefinitionToJson(this);
}

/// Complete filter schema definition.
@JsonSerializable()
class FilterSchema {
  /// JSON Schema reference (optional).
  @JsonKey(name: r'$schema')
  final String? schema;

  /// Unique identifier for this filter.
  final String id;

  /// Schema version.
  final String version;

  /// Display name.
  final String name;

  /// Short one-line description of what this filter does.
  final String? description;

  /// Longer prose shown above the filter's options: what the filter does and
  /// when you'd reach for it. Paragraphs are separated by a blank line.
  /// Falls back to [description] in the UI when absent.
  final String? longDescription;

  /// Category for grouping (e.g., "cleanup", "enhancement").
  final String? category;

  /// Icon name (Material icon).
  final String? icon;

  /// Sort order in filter list.
  final int order;

  /// Highest per-component bit depth this filter processes natively. When set
  /// and the source exceeds it, the worker down-converts to this depth for the
  /// pass (a lossy round-trip), so the UI warns of reduced precision. Null (the
  /// default) means the filter handles the source's bit depth without loss.
  final int? maxBitDepth;

  /// Dependencies required by this filter.
  final FilterDependencies? dependencies;

  /// Available methods for this filter.
  final List<MethodDefinition> methods;

  /// Parameter definitions.
  final Map<String, ParameterDefinition> parameters;

  /// Parameter preset selectors - dropdowns that set one or more parameters.
  final Map<String, ParameterPreset>? parameterPresets;

  /// Preset configurations (for QTGMC-style quality presets).
  final Map<String, Map<String, dynamic>>? presets;

  /// UI layout configuration.
  final UiLayout? ui;

  /// Code generation configuration.
  final CodeTemplate? codeTemplate;

  /// The VapourSynth calls this filter makes, when it is not simply the
  /// selected method's own `function`. Declared only by the composite passes —
  /// see [ImplementationEntry].
  final List<ImplementationEntry>? implementation;

  /// Source of this schema: "builtin", "user", "community".
  @JsonKey(includeFromJson: false, includeToJson: false)
  String source;

  FilterSchema({
    this.schema,
    required this.id,
    required this.version,
    required this.name,
    this.description,
    this.longDescription,
    this.category,
    this.icon,
    this.order = 0,
    this.maxBitDepth,
    this.dependencies,
    required this.methods,
    required this.parameters,
    this.parameterPresets,
    this.presets,
    this.ui,
    this.codeTemplate,
    this.implementation,
    this.source = 'builtin',
  });

  /// Get the default method.
  MethodDefinition get defaultMethod => methods.first;

  /// Get a method by ID.
  MethodDefinition? getMethod(String id) {
    try {
      return methods.firstWhere((m) => m.id == id);
    } catch (_) {
      return null;
    }
  }

  /// The methods to offer in the UI.
  ///
  /// In simple mode ([showAdvanced] false) advanced-only methods are dropped —
  /// except [selectedId], which is always kept. Hiding a value the user has
  /// already chosen would both lie about what the pipeline is doing and, in a
  /// dropdown, crash on a value that isn't among its items.
  ///
  /// Order is preserved, so the schema decides what a regular user sees first.
  List<MethodDefinition> visibleMethods({
    required bool showAdvanced,
    String? selectedId,
  }) {
    if (showAdvanced) return methods;
    return methods
        .where((m) => !m.advancedOnly || m.id == selectedId)
        .toList();
  }

  /// Whether any method is advanced-only, so the UI knows there is more to
  /// reveal even when every parameter is basic.
  bool get hasAdvancedMethods => methods.any((m) => m.advancedOnly);

  /// Get default values for all parameters.
  /// Optional parameters are excluded (they start as null/disabled).
  Map<String, dynamic> getDefaults() {
    final defaults = <String, dynamic>{};
    for (final entry in parameters.entries) {
      // Skip optional parameters - they start as disabled (null)
      if (entry.value.optional == true) continue;
      defaults[entry.key] = entry.value.defaultValue;
    }
    return defaults;
  }

  /// Validate parameter values against schema.
  List<String> validate(Map<String, dynamic> values) {
    final errors = <String>[];
    for (final entry in values.entries) {
      final param = parameters[entry.key];
      if (param == null) {
        errors.add('Unknown parameter: ${entry.key}');
        continue;
      }
      if (!param.isValidValue(entry.value)) {
        errors.add('Invalid value for ${entry.key}: ${entry.value}');
      }
    }
    return errors;
  }

  factory FilterSchema.fromJson(Map<String, dynamic> json) =>
      _$FilterSchemaFromJson(json);
  Map<String, dynamic> toJson() => _$FilterSchemaToJson(this);
}

File: worker/src/filter_schema.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Type of parameter value.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ParameterType {
    Boolean,
    Integer,
    Number,
    String,
    #[serde(rename = "enum")]
    Enum,
}

/// Type of UI widget to render for a parameter.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum WidgetType {
    Slider,
    Dropdown,
    Checkbox,
    Textfield,
    Number,
}

/// VapourSynth-specific parameter configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VapourSynthConfig {
    /// The parameter name in VapourSynth (may differ from schema name).
    pub name: String,
}

/// UI configuration for a parameter.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ParameterUiConfig {
    /// Display label for the parameter.
    pub label: Option<String>,

    /// Description/tooltip for the parameter.
    pub description: Option<String>,

    /// Type of widget to render.
    pub widget: Option<WidgetType>,

    /// Decimal precision for number display.
    pub precision: Option<i32>,

    /// Whether this parameter is hidden from the UI.
    pub hidden: Option<bool>,

    /// Condition for when this parameter is visible.
    #[serde(rename = "visibleWhen")]
    pub visible_when: Option<HashMap<String, serde_json::Value>>,
}

/// Definition of a single parameter.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParameterDefinition {
    /// Type of this parameter.
    #[serde(rename = "type")]
    pub param_type: ParameterType,

    /// Default value (type depends on param_type).
    #[serde(rename = "default")]
    pub default_value: serde_json::Value,

    /// Minimum value (for number/integer types).
    pub min: Option<f64>,

    /// Maximum value (for number/integer types).
    pub max: Option<f64>,

    /// Step size for sliders.
    pub step: Option<f64>,

    /// Available options (for enum type).
    pub options: Option<Vec<String>>,

    /// VapourSynth-specific configuration.
    pub vapoursynth: Option<VapourSynthConfig>,

    /// UI configuration.
    pub ui: Option<ParameterUiConfig>,
}

impl ParameterDefinition {
    /// Get the VapourSynth parameter name.
    pub fn get_vs_name<'a>(&'a self, schema_name: &'a str) -> &'a str {
        self.vapoursynth
            .as_ref()
            .map(|vs| vs.name.as_str())
            .unwrap_or(schema_name)
    }

    /// Check if a value is valid for this parameter.
    pub fn is_valid_value(&self, value: &serde_json::Value) -> bool {
        match (&self.param_type, value) {
            (ParameterType::Boolean, serde_json::Value::Bool(_)) => true,
            (ParameterType::Integer, serde_json::Value::Number(n)) => {
                if !n.is_i64() {
                    return false;
                }
                let v = n.as_i64().unwrap();
                if let Some(min) = self.min {
                    if (v as f64) < min {
                        return false;
                    }
                }
                if let Some(max) = self.max {
                    if (v as f64) > max {
                        return false;
                    }
                }
                true
            }
            (ParameterType::Number, serde_json::Value::Number(n)) => {
                if let Some(v) = n.as_f64() {
                    if let Some(min) = self.min {
                        if v < min {
                            return false;
                        }
                    }
                    if let Some(max) = self.max {
                        if v > max {
                            return false;
                        }
                    }
                    true
                } else {
                    false
                }
            }
            (ParameterType::String, serde_json::Value::String(_)) => true,
            (ParameterType::Enum, serde_json::Value::String(s)) => {
                self.options.as_ref().map(|opts| opts.contains(s)).unwrap_or(true)
            }
            (_, serde_json::Value::Null) => true,
            _ => false,
        }
    }
}

/// Definition of a filter method (e.g., DeHalo_alpha, YAHR).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MethodDefinition {
    /// Unique identifier for this method.
    pub id: String,

    /// Display name.
    pub name: String,

    /// Description of what this method does.
    pub description: Option<String>,

    /// VapourSynth function to call (e.g., "haf.DeHalo_alpha").
    pub function: String,

    /// List of parameter IDs that this method uses.
    pub parameters: Vec<String>,
}

/// UI section grouping parameters together.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UiSection {
    /// Section title.
    pub title: String,

    /// Parameter IDs in this section.
    pub parameters: Vec<String>,

    /// Whether the section is expanded by default.
    #[serde(default = "default_true")]
    pub expanded: bool,

    /// Whether this section only appears in advanced mode.
    #[serde(default)]
    pub advanced_only: bool,
}

fn default_true() -> bool {
    true
}

/// UI layout configuration.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UiLayout {
    /// Sections to organize parameters.
    pub sections: Option<Vec<UiSection>>,
}

/// Dependencies required by a filter.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct FilterDependencies {
    /// Python packages required (e.g., ["havsfunc", "mvsfunc"]).
    pub plugins: Option<Vec<String>>,

    /// VapourSynth plugins required (e.g., ["libmvtools.dll"]).
    #[serde(rename = "vs_plugins")]
    pub vs_plugins: Option<Vec<String>>,

    /// Optional plugins that enable additional features.
    pub optional: Option<Vec<String>>,
}

/// Code generation configuration.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct CodeTemplate {
    /// Import statements needed.
    pub imports: Option<Vec<String>>,

    /// How to generate code: "method" (use method.function) or custom template.
    pub generate: Option<String>,

    /// Custom code body template.
    pub body: Option<String>,
}

/// Complete filter schema definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FilterSchema {
    /// JSON Schema reference (optional).
    #[serde(rename = "$schema")]
    pub schema: Option<String>,

    /// Unique identifier for this filter.
    pub id: String,

    /// Schema version.
    pub version: String,

    /// Display name.
    pub name: String,

    /// Description of what this filter does.
    pub description: Option<String>,

    /// Category for grouping.
    pub category: Option<String>,

    /// Icon name.
    pub icon: Option<String>,

    /// Sort order in filter list.
    #[serde(default)]
    pub order: i32,

    /// Dependencies required by this filter.
    pub dependencies: Option<FilterDependencies>,

    /// Available methods for this filter.
    pub methods: Vec<MethodDefinition>,

    /// Parameter definitions.
    pub parameters: HashMap<String, ParameterDefinition>,

    /// Preset configurations.
    pub presets: Option<HashMap<String, HashMap<String, serde_json::Value>>>,

    /// UI layout configuration.
    pub ui: Option<UiLayout>,

    /// Code generation configuration.
    #[serde(rename = "codeTemplate")]
    pub code_template: Option<CodeTemplate>,

    /// Source of this schema (not serialized).
    #[serde(skip)]
    pub source: String,
}

impl FilterSchema {
    /// Get the default method.
    pub fn default_method(&self) -> Option<&MethodDefinition> {
        self.methods.first()
    }

    /// Get a method by ID.
    pub fn get_method(&self, id: &str) -> Option<&MethodDefinition> {
        self.methods.iter().find(|m| m.id == id)
    }

    /// Get default values for all parameters.
    pub fn get_defaults(&self) -> HashMap<String, serde_json::Value> {
        self.parameters
            .iter()
            .map(|(k, v)| (k.clone(), v.default_value.clone()))
            .collect()
    }

    /// Validate parameter values against schema.
    pub fn validate(&self, values: &HashMap<String, serde_json::Value>) -> Vec<String> {
        let mut errors = Vec::new();

        for (key, value) in values {
            match self.parameters.get(key) {
                Some(param) => {
                    if !param.is_valid_value(value) {
                        errors.push(format!("Invalid value for {}: {:?}", key, value));
                    }
                }
                None => {
                    errors.push(format!("Unknown parameter: {}", key));
                }
            }
        }

        errors
    }
}

/// Dynamic parameter container for schema-based filters.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DynamicParameters {
    /// The filter ID this belongs to.
    pub filter_id: String,

    /// Whether this filter pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Parameter values.
    #[serde(default)]
    pub values: HashMap<String, serde_json::Value>,
}

impl DynamicParameters {
    /// Create with default values from a schema.
    pub fn from_schema(schema: &FilterSchema, enabled: bool) -> Self {
        Self {
            filter_id: schema.id.clone(),
            enabled,
            values: schema.get_defaults(),
        }
    }

    /// Get a boolean parameter.
    pub fn get_bool(&self, key: &str) -> Option<bool> {
        self.values.get(key).and_then(|v| v.as_bool())
    }

    /// Get an integer parameter.
    pub fn get_int(&self, key: &str) -> Option<i64> {
        self.values.get(key).and_then(|v| v.as_i64())
    }

    /// Get a float parameter.
    pub fn get_float(&self, key: &str) -> Option<f64> {
        self.values.get(key).and_then(|v| v.as_f64())
    }

    /// Get a string parameter.
    pub fn get_string(&self, key: &str) -> Option<&str> {
        self.values.get(key).and_then(|v| v.as_str())
    }

    /// Get the currently selected method ID.
    pub fn method(&self) -> Option<&str> {
        self.get_string("method")
    }
}

/// Container for all dynamic filter parameters in a pipeline.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct DynamicPipeline {
    /// Map of filter ID to parameters.
    #[serde(default)]
    pub filters: HashMap<String, DynamicParameters>,
}

impl DynamicPipeline {
    /// Get parameters for a specific filter.
    pub fn get(&self, filter_id: &str) -> Option<&DynamicParameters> {
        self.filters.get(filter_id)
    }

    /// Check if a filter is enabled.
    pub fn is_enabled(&self, filter_id: &str) -> bool {
        self.filters.get(filter_id).map(|p| p.enabled).unwrap_or(false)
    }

    /// Get list of enabled filter IDs.
    pub fn enabled_filter_ids(&self) -> Vec<&str> {
        self.filters
            .iter()
            .filter(|(_, p)| p.enabled)
            .map(|(id, _)| id.as_str())
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_filter_schema() {
        let json = r#"{
            "id": "dehalo",
            "version": "1.0.0",
            "name": "Dehalo",
            "order": 4,
            "methods": [
                {
                    "id": "dehalo_alpha",
                    "name": "DeHalo Alpha",
                    "function": "haf.DeHalo_alpha",
                    "parameters": ["rx", "ry"]
                }
            ],
            "parameters": {
                "method": {
                    "type": "enum",
                    "default": "dehalo_alpha",
                    "options": ["dehalo_alpha", "yahr"]
                },
                "rx": {
                    "type": "number",
                    "default": 2.0,
                    "min": 1.0,
                    "max": 3.0
                }
            }
        }"#;

        let schema: FilterSchema = serde_json::from_str(json).unwrap();
        assert_eq!(schema.id, "dehalo");
        assert_eq!(schema.methods.len(), 1);
        assert_eq!(schema.parameters.len(), 2);
    }

    #[test]
    fn test_parameter_validation() {
        let param = ParameterDefinition {
            param_type: ParameterType::Number,
            default_value: serde_json::json!(2.0),
            min: Some(1.0),
            max: Some(3.0),
            step: Some(0.1),
            options: None,
            vapoursynth: None,
            ui: None,
        };

        assert!(param.is_valid_value(&serde_json::json!(2.0)));
        assert!(param.is_valid_value(&serde_json::json!(1.0)));
        assert!(param.is_valid_value(&serde_json::json!(3.0)));
        assert!(!param.is_valid_value(&serde_json::json!(0.5)));
        assert!(!param.is_valid_value(&serde_json::json!(3.5)));
    }
}
